### PR TITLE
Cpu strided complex unaryop

### PR DESCRIPTION
**Pull request by dylanbespalko**
@ezyang 
In-tree changes to pytorch to support complex numbers are being submitted here.
Out-of-tree support for complex numbers is here: [pytorch-cpu-strided-complex extension](https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex)

Changes in this PR.

- [x]  Added complex support of UnaryOp kernels.

I have fixed several problems that prevented the complex number kernels from re-using existing code:
1. Added fake AVX support that allows complex kernels to call `cpu_kernel_vec` instead of `cpu_kernel`.
  - `aten/src/ATen/cpu/vec256/vec256_complex_double.h`
  - `aten/src/ATen/cpu/vec256/vec256_complex_double.h`
  - Code was taken from @Roger-luo complex number repo.
  - I also added `real`, `imag` methods to all other data types and `vec256_base.h`
2. Added `abs`, `real`, `imag` utility functions that operate on complex dtypes and act as no-ops for other data types. 
  - 'aten/src/ATen/native/cpu/ZMath.h` contains these utility functions.

One final problem exists: `std::complex` does not support several compare operators. Would it be acceptable to replace occurrences of `a > b` with `gt(a, b)`, where `gt` is implemented similar to `real_impl` and `imag_impl` in `ZMath.h`?

### CODE REVIEW

**Comment by ezyang** [comment]
> One final problem exists: std::complex does not support several compare operators. Would it be acceptable to replace occurrences of a > b with gt(a, b), where gt is implemented similar to real_impl and imag_impl in ZMath.h?

I have a broader question to ask: how exactly are you defining comparison on complex numbers? This is one of those cases where I feel we should NOT provide an implementation.

**Comment by ezyang** [comment]
@gchanan if you could look at the dispatch macro bits

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -131,6 +131,40 @@
     CPU: _abs_out_cpu
     CUDA: _abs_out_cuda
 
+- func: real(Tensor self) -> Tensor
+  variants: function, method
+  named_guard: False
+
+- func: real_(Tensor(a!) self) -> Tensor(a!)
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml]: `real_` changes the dtype of a tensor, so I'm not sure how you can actually do it inplace...

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 2]
```diff
@@ -82,12 +101,21 @@ static void bitwise_not_kernel(TensorIterator& iter) {
...
-        [=](scalar_t a) -> scalar_t { return a - std::trunc(a); },
-        [=](Vec256<scalar_t> a) { return a.frac(); });
-  });
+  if (isComplexType(iter.dtype())) {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "frac_cpu", [&]() {
+      cpu_kernel(
+          iter,
+          [=](scalar_t a) -> scalar_t {
+            return a - scalar_t(std::trunc(std::real(a)), std::trunc(std::imag(a))); });
+    });
```
Comment (ezyang) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp]: lol, I'm not sure if this is ever actually a useful operation on complex numbers, but sure

File: aten/src/ATen/cpu/vec256/vec256_complex_double.h [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+public:
+  using value_type = std::complex<double>;
+  static constexpr int size() {
+    return 2;
+  }
+  Vec256() {}
+  Vec256(std::complex<double> val) {
+    for (int i = 0; i != size(); i++) {
+      values[i] = val;
+    }
```
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256_complex_double.h]: These operations haven't been vectorized at all! The loops aren't even unrolled!

In other words, what it seems like what you've implemented here is not the equivalent of `vec256_float.h` (which is all vectorized) but `vec256_base.h` (which provides non-vectorized implementations of all operations to default back to.) So, I have two questions:

1. Can we rename the header so this is clear?
2. Can we deduplicate `vec256_complex_{double,float}`, since if these are non-vectorized, the implementations should basically be the same. This is similar to how `base` is only written once.

**Review by ezyang — CHANGES_REQUESTED** [review]
This looks pretty good. However, inplace real/imag, and non-vectorized header, are the show stoppers for landing.

**Comment by dylanbespalko** [comment]
> @gchanan if you could look at the dispatch macro bits

@gchanan: I'm trying NOT to support half precision complex numbers right now, but I found that `eps`, `min`, `max`, and `tiny` require that I implement `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND_HALF`. This is the only place in the code where this dispatch macro is needed.

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 2 above)
Comment (dylanbespalko) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp]: I agree, I will remove it. I was using it early on because it was an easy function that doesn't all `cpu_vec`.

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (dylanbespalko) [reply aten/src/ATen/native/native_functions.yaml]: These should absolutely NOT be inplace. Thanks for catching this.

**Comment by dylanbespalko** [comment]
@ezyang 
 
I have renamed the unvectorized code to `vec256_complex_double_no_vec.h` and `vec256_complex_float_no_vec.h`. I will likely combine complex float and complex double after I have a chance to implement some AVX vectorizations.

I have removed the inplace calls to `real_` and `imag_`.

**I have to run CI again to fix a workaround for a bug in `Glibc2.23` called `DL_RUNTIME_BUG` that didn't work for complex numbers.**

Finally, I have made changes based on the recommendations that you made above. I plan to add comparison ops next, however I will only add the operators that are supported by std::complex until I have a better sense of how kernels like `torch.clamp` become vectorized.

I will contact you when the CI is passing.

**Comment by dylanbespalko** [comment]
> I have a broader question to ask: how exactly are you defining comparison on complex numbers? This is one of those cases where I feel we should NOT provide an implementation.

I would define a `a > b` as `abs(a) > abs(b)` where `abs() = sqrt(a.real()**2 + a.imag()**2)`.  I work with complex numbers every day and I have never seen another interpretation in real life, however I have noticed that numpy defines `a > b` as `a.real() > b.real()`. Maybe there is ambiguity???

In a library like Pytorch I could see this issue propagating when people define compounded operations like `torch.clamp`. Anyways, I won't implement it until I have a better understanding of how an operation like `torch.clamp` is vectorized for complex numbers.

**Comment by ezyang** [comment]
> I would define a a > b as abs(a) > abs(b) where abs() = sqrt(a.real()**2 + a.imag()**2). I work with complex numbers every day and I have never seen another interpretation in real life, however I have noticed that numpy defines a > b as a.real() > b.real()

This is kind of what I have expected, but Wikipedia says that this doesn't follow some laws (which I also believe). It is surprising that Numpy only compares reals haha.

**Comment by dylanbespalko** [comment]
Still trying to figure out what went wrong with the CI. I'm going to have to step though the changes tomorrow.

**Comment by vadimkantorov** [comment]
This should fix https://github.com/pytorch/pytorch/issues/22764. Also sometimes abs squared is needed (which can save sqrt + square ops and bring more numerical stability). `abs(squared = True)` would be nice, but maybe not so generic. In general, Lp norm to p'th power is not unusual. 

Is abs on complex type faster than norm(dim = -1) on a float type? If yes, maybe abs should be called underneath in norm if last dimension of a float tensor is 2.

**Comment by dylanbespalko** [comment]
@vadimkantorov 

Yes, I agree. It seems the reason why some complex operators are not implemented in std::complex is that you will find other ways to optimize the computation. Note: this change does not support vectorization yet, but I will work on that as soon as I figure out this CI problem.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/Dispatch.h [hunk 4]
```diff
@@ -139,6 +139,21 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
...
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Double, double, __VA_ARGS__)                                          \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Float, float, __VA_ARGS__)                                            \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::ComplexDouble, std::complex<double>, __VA_ARGS__)                     \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::ComplexFloat, std::complex<float>, __VA_ARGS__)                       \
+      default:                                                                                                   \
+        AT_ERROR(#NAME, " not implemented for '", toString(_st), "'");                                           \
+    }                                                                                                            \
+  }()
+
+ #define AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND_HALF(TYPE, NAME, ...)                                        \
```
[the remark this answers is not in the document]
Comment (dylanbespalko) [reply aten/src/ATen/Dispatch.h]: This sounds fine. I'll make the change.

File: aten/src/ATen/native/UnaryOps.cpp [hunk 5]
```diff
@@ -239,9 +239,9 @@ inline void propagate_names_if_namedtensor_enabled(Tensor& result, const Tensor&
     return at::op##_out(self, self);                                   \
   }                                                                    \
   Tensor& _##op##_out_##prefix(Tensor& result, const Tensor& self) {   \
-    checkBackend(#op, result, Backend::device);                        \
     auto iter = TensorIterator::unary_op(result, self,                 \
       /*check_mem_overlap=*/true);                                     \
+    AT_ASSERT(iter.device_type() == DeviceType::CPU);                  \
```
[the remark this answers is not in the document]
Comment (dylanbespalko) [reply aten/src/ATen/native/UnaryOps.cpp]: Ok, I think I get it.
`checkDeviceType()` is only checking device.
`checkBackend()` is checking both device and layout.

**Comment by gchanan** [comment]
@VitalyFedyunin can you comment on the vec256 changes?

**Comment by dylanbespalko** [comment]
@VitalyFedyunin, @gchanan,

I wouldn't spend too much time looking at my Vec256 changes. @ezyang has given me some help on that and I should have a Vec256 complex number implementation soon.

**Comment by VitalyFedyunin** [comment]
> @VitalyFedyunin, @gchanan,
> 
> I wouldn't spend too much time looking at my Vec256 changes. @ezyang has given me some help on that and I should have a Vec256 complex number implementation soon.

Got it, add me when you feel ready to receive Vec256 feedback. But unless you actually start using avx2 instructions it is kind of over complicated to add two more classes when you can just make cpu_kernel_vec to ignore vectorization in case of complex types.
